src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,153 @@
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", is out of bound!");
+        }
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> temp = head;
+            Node<T> newNode = new Node<>(null, value, temp);
+            head = newNode;
+            if (temp != null) {
+                temp.prev = newNode;
+            }
+            size++;
+        } else {
+            Node<T> temp = getNode(index);
+            Node<T> newNode = new Node<>(temp.prev, value, temp);
+            temp.prev.next = newNode;
+            temp.prev = newNode;
+            size++;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T elem : list) {
+            add(elem);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> temp = getNode(index);
+        T oldValue = temp.value;
+        temp.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 0) {
+            throw new NoSuchElementException("No such element: " + object);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value == object || current.value != null && current.value.equals(object)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", is out of bound!");
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private T unlink(Node<T> node) {
+        if (node.equals(head)) {
+            head = head.next;
+            if (size == 1) {
+                tail = null;
+            }
+        } else if (node.equals(tail)) {
```

Comment:
you use a method from the Object class that compares references as `==`, simplify it